### PR TITLE
[RFC] - Config is code

### PR DESCRIPTION
**Pull request by felipemello1**
This PR attempts to showcase the different styles we could use for our config system. A push here is if we can leave the .yaml and use .py instead, under the reasoning that "config is code". 

Python can enable a) better type safety, b) import directly instead of using strings, and c) more flexibility for the user to define custom code within the config.

On the other hand, yaml can be easier to read and hydra provides some nice freebies.

Must have:
- Lazy instantiation: Currently we don't have any way to instantiate a function through configs, e.g. 'loss: my_loss_fn'.
- Handle complex patterns, e.g. nesting

Nice to have:
- Composition: Our code is naturally fragmented (generator, rewards, etc). Touching one config without touching the others is a relevant feature.
- Ability to code in the config, e.g. if statements, sum values, check conditions, etc.
- CLI overrides

Proposed options:
1. OmegaConfg / Hydra
2. Fiddle (cfg option by google)
3. Factory + dataclasses
4. Plain python with partials
5. Plain python with dictionary
TODO: toml

I would like to hear what others think about the options, if we should stick with yaml and fully explore hydra/omegaconf or change to .py system.

### CODE REVIEW

File: brainstorming/configs/config_partial.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    # =========================================================================
+    cfg2 = llama3_2_1b_full()
+
+    # PATTERN 1: Simple Component Instantiation
+    tokenizer2 = cfg2["tokenizer"](path="/new/path")
+
+    # PATTERN 2: Component with Nested Instantiation
+    inner_partial = cfg2["model"].keywords["attn_config"]
+    inner_partial.keywords["num_heads"] = 64
+    model2 = cfg2["model"]()
```
Comment (tianyu-l) [inline brainstorming/configs/config_partial.py:93]: definition part is OK, but its consumption is confusing when it comes to nested init

File: brainstorming/configs/config_dicts.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    # =========================================================================
+    def llama3_2_1b_large_lr():
+        """Variant with larger learning rate."""
+        base = llama3_2_1b_full()
+        base["optimizer"]["kwargs"]["lr"] = 1e-3
+        base["model"]["kwargs"]["attn_config"]["kwargs"]["num_heads"] = 64
+        return base
+
+    cfg_variant = llama3_2_1b_large_lr()
+    attn_config_variant = cfg_variant["model"]["kwargs"]["attn_config"]["cls"](
```
Comment (tianyu-l) [inline brainstorming/configs/config_dicts.py:124]: seems too flexible, hard to read, and error-prone

File: brainstorming/configs/config_dataclasses.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Dataclass config with inner Config classes.
```
Comment (tianyu-l) [inline brainstorming/configs/config_dataclasses.py:8]: can we get free cli override with tyro?
https://github.com/pytorch/torchtitan/blob/2ea6197b957936bdd4941e59a000cf31987a3184/torchtitan/config/manager.py#L56

File: brainstorming/configs/config_hydra.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    tokenizer2 = instantiate(cfg2.tokenizer)
+
+    # PATTERN 2: Component with Nested Instantiation
+    cfg2.model.attn_config.num_heads = 64
+    model2 = instantiate(cfg2.model)
+
+    # PATTERN 3: Component Needing Runtime Args (Partial)
+    cfg2.optimizer.lr = 1e-4
+    optimizer_partial2 = instantiate(cfg2.optimizer)
+    optimizer2 = optimizer_partial2(params=model2.parameters())
```
Comment (tianyu-l) [inline brainstorming/configs/config_hydra.py:69]: This looks a bit confusing.

File: brainstorming/configs/config_hydra.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+"""
+Config using Hydra (YAML-based lazy instantiation).
+
+Pros:
+- YAML syntax (human-readable)
+- Native composition of yamls
+- Lazy instantiation via hydra.utils.instantiate
+- Command-line override for free (--optimizer.lr=1e-4)
+
+Cons:
```
Comment (tianyu-l) [inline brainstorming/configs/config_hydra.py:16]: plus learning curve, `compose`, `instantiate`, etc.

File: brainstorming/configs/config_dataclasses.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    """Plain dataclass for non-instantiated config block (PATTERN 4)."""
+
+    batch_size: int = 4
+    shuffle: bool = True
+
+
+def llama3_2_1b_full():
+    output_dir = "/tmp/torchtune/llama3_2_1B/full"
+
+    return {
```
Comment (tianyu-l) [inline brainstorming/configs/config_dataclasses.py:57]: For this one and the fiddle one, I guess you can take this to extreme and make everything a (data)class, e.g. `TrainerWithConfig` class can have model, optimizer, data loader, etc. What's your thought?

**Review by tianyu-l — COMMENTED** [review]
Thanks for putting up these candidates! Left some initial impressions, might not be accurate.

File: brainstorming/configs/config_dicts.py (hunk 2 above)
Comment (felipemello1) [reply brainstorming/configs/config_dicts.py:124]: agreed, its horrible. I put it here to make a contrast

File: brainstorming/configs/config_hydra.py (hunk 4 above)
Comment (felipemello1) [reply brainstorming/configs/config_hydra.py:69]: agreed, perhaps there is a better way? Not sure if i can pass the params already in instantiate.

The spirit here is that i can actually do once:
```instantiate(cfg)```, and everything gets instantiated already at the very start, but when i call cfg.optimizer, i will get a `partial`.

In this example it feels weird because instead of doing `instantiate(config)```, i am instantiating each arg of the config individually

File: brainstorming/configs/config_dataclasses.py (hunk 6 above)
Comment (felipemello1) [reply brainstorming/configs/config_dataclasses.py:57]: My issue with this pattern of `TrainerWithConfig.config()` is that its too opinionated and impacts the entire codebase. 3rd party utilities, or local code that dont need to be a class, must be represented as such.

Between the two, i personally prefer fiddle.

But when i read `config_fiddle.py` and i read `baseline.yaml`, to my eyes,` baseline.yaml` is easier to parse and understand. Perhaps because its a simple config? I can imagine cases where using python can be handy.

On the composability side, if you look at `baseline_different_bsz.yaml`, it seems easier to abstract away experimentation from infra args.

TDLR
1. despite all of the hate, IMO .yaml + OmegaConf and/or Hydra is the lesser of all evils.
2. If we want to use .py, fiddle seems the easiest
3. dataclasses.config pattern are the safest, but impact the entire code base and is harder to read

File: brainstorming/configs/config_dataclasses.py (hunk 3 above)
Comment (felipemello1) [reply brainstorming/configs/config_dataclasses.py:8]: that looks nice, didnt know about it

File: brainstorming/configs/config_hydra.py (hunk 5 above)
Comment (felipemello1) [reply brainstorming/configs/config_hydra.py:16]: in practice, i think that people would only use instantiate, and we can even get rid of this and instantiate the whole config once at the start. Partials can then be called where needed. e.g.

```
def main(path_cfg:str):
	cfg = instantiate(load_cfg(path_cfg))
	
    model = cfg.model
	optimizer = cfg.optimizer(params=model.param)
```

**Comment by felipemello1** [comment]
After some research, it seems fair to say that:
1. OmegaConf is battle tested and the choice of most frameworks (amaia, torchtune, verl, litgpt, lingua, axolotl/trl use yml)
2. Frameworks that need dynamism will have to figure out some way to do (lazy) instantiation. This manifests in the following ways:

	a. **Hardcoded**: users don't have flexibility changing any module. They can only pass parameters.

	b. **Registry**: Users can pick from a registry. They cant pass 3rd party or custom modules. Example from [Amaia](https://github.com/fairinternal/amaia/blob/1379fd403c798b1f46ea8dfd15f8d945f1fa2145/apps/rl/envs/config.py#L45) and [axolotl](https://github.com/axolotl-ai-cloud/axolotl/blob/c37decb073dcfa3538f96bf8a9f689ca5b76befd/src/axolotl/utils/schemas/enums.py#L73).

	c. **Custom utilities to instantiate**: Library defines their own utilities. Example is [TorchTune](https://github.com/meta-pytorch/torchtune/blob/67ab86b94de9e7ac7dd9850113ebe69e2bbd307c/torchtune/config/_instantiate.py#L69) with `_component_` and [litgpt](https://github.com/Lightning-AI/litgpt/blob/062fff2f96f27416cc866eefb32352955234a332/config_hub/finetune/gemma-7b/lora.yaml#L48C3-L48C13) with `class_path`.

	d. **Hydra**: Example is [VeRL](https://github.com/volcengine/verl/blob/2deda718f03cb5b905dd30096ee55cb622905417/verl/utils/config.py#L50)

3. VeRL shows that using Hydra to do it can have minimal impact, without having to reinvent the wheel. Only 2 places are impacted in the codebase: i) [config utils](https://github.com/volcengine/verl/blob/2deda718f03cb5b905dd30096ee55cb622905417/verl/utils/config.py#L50); ii) [decorator in main.py](https://github.com/volcengine/verl/blob/2deda718f03cb5b905dd30096ee55cb622905417/verl/trainer/sft_trainer.py#L372) to get the freebies from hydra

Therefore, it seems reasonable for me to conclude that:
i) We should stick with yaml + OmegaConf;
ii) If we want to make users lives easier, we want to allow them to instantiate custom modules;
iii) Using hydra can be designed in a way that has minimal impact in the codebase and is way easier than managing our own custom utilities
